usersession/agent/rest_api.go

```diff
@@ -334,3 +355,218 @@ func postRefreshFinishedNotification(c *Command, r *http.Request) Response {
+	extraParams := make(map[string]dbus.Variant)
+	if destroy {
+		err = obj.Call("io.snapcraft.SnapDesktopIntegration.ApplicationRefreshCompleted", 0, snapName, extraParams).Store()
+	} else {
+		err = obj.Call("io.snapcraft.SnapDesktopIntegration.ApplicationIsBeingRefreshed", 0, snapName, "", extraParams).Store()
+	}
+	if err != nil {
+		logger.Noticef("unable to successfully call io.snapcraft.SnapDesktopIntegration: %v", err)
+		return false
+	}
+	return true
+}
+
+func postBeginDeferredRefreshNotification(c *Command, r *http.Request) Response {
+	if ok, resp := validateJSONRequest(r); !ok {
+		return resp
+	}
+
+	decoder := json.NewDecoder(r.Body)
+
+	// pendingSnapRefreshInfo holds information about pending snap refresh provided by snapd.
+	var refreshInfo client.DeferredSnapRefreshInfo
+	if err := decoder.Decode(&refreshInfo); err != nil {
+		return BadRequest("cannot decode request body into pending snap deferred refresh info: %v", err)
+	}
+
+	if tryNotifyRefreshToSnapDesktopIntegration(refreshInfo.InstanceName, false) {
+		return SyncResponse(&resp{
+			Type:   ResponseTypeError,
+			Status: 200,
+			Result: &errorResult{
+				Message: "Ok",
+			},
+		})
+	}
+
+	// Note that since the connection is shared, we are not closing it.
+	if c.s.bus == nil {
+		return SyncResponse(&resp{
+			Type:   ResponseTypeError,
+			Status: 500,
+			Result: &errorResult{
+				Message: "cannot connect to the session bus",
+			},
+		})
+	}
+
+	// TODO: this message needs to be crafted better as it's the only thing guaranteed to be delivered.
+	summary := fmt.Sprintf(i18n.G("Updating '%q' snap"), refreshInfo.InstanceName)
+	body := i18n.G("Please wait before opening it.")
+	var icon string
+	var hints []notification.Hint
+
+	hints = append(hints, notification.WithUrgency(notification.NormalUrgency))
+	// The notification is provided by snapd session agent.
+	hints = append(hints, notification.WithDesktopEntry("io.snapcraft.SessionAgent"))
+	// But if we have a desktop file of the busy application, use that apps's icon.
+	if refreshInfo.AppDesktopEntry != "" {
+		parser := goconfigparser.New()
+		desktopFilePath := filepath.Join(dirs.SnapDesktopFilesDir, refreshInfo.AppDesktopEntry+".desktop")
+		if err := parser.ReadFile(desktopFilePath); err == nil {
+			icon, _ = parser.Get("Desktop Entry", "Icon")
+		}
+	}
+
+	msg := &notification.Message{
+		AppName: refreshInfo.AppName,
+		Title:   summary,
+		Icon:    icon,
+		Body:    body,
+		Hints:   hints,
+	}
+
+	// TODO: silently ignore error returned when the notification server does not exist.
+	// TODO: track returned notification ID and respond to actions, if supported.
+	if err := c.s.notificationMgr.SendNotification(notification.ID(refreshInfo.InstanceName), msg); err != nil {
+		return SyncResponse(&resp{
+			Type:   ResponseTypeError,
+			Status: 500,
+			Result: &errorResult{
+				Message: fmt.Sprintf("cannot send notification message: %v", err),
+			},
+		})
+	}
+	return SyncResponse(nil)
```

Comment:
I see almost identical code here and in `postFinishDeferredRefreshNotification` could some of this be reused?

Reply:
Done.